File: .NET/Microsoft.Recognizers.Text.DateTime/Utilities/Token.cs
﻿// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.

using System.Collections.Generic;
using System.Linq;
using System.Text.RegularExpressions;

using Microsoft.Recognizers.Text.InternalCache;

namespace Microsoft.Recognizers.Text.DateTime
{
    public class Token : ICloneableType<Token>
    {

        public Token(int s, int e, Metadata metadata = null)
        {
            Start = s;
            End = e;
            Metadata = metadata;
        }

        public int Start { get; }

        public int End { get; }

        public Metadata Metadata { get; }

        public int Length
        {
            get
            {
                if (End < Start)
                {
                    return 0;
                }

                return End - Start;
            }
        }

        public static List<Token> GetTokenFromRegex(Regex regex, string text)
        {
            var ret = new List<Token>();

            if (regex == null)
            {
                return ret;
            }

            var matches = regex.Matches(text);
            foreach (Match match in matches)
            {
                ret.Add(new Token(match.Index, match.Index + match.Length));
            }

            return ret;
        }

        public static List<ExtractResult> MergeAllTokens(List<Token> tokens, string text, string extractorName)
        {
            var ret = new List<ExtractResult>();

            tokens = tokens.OrderBy(s => s.Start).ThenByDescending(s => s.Length).ToList();

            var mergedTokens = new List<Token>();

            foreach (var token in tokens)
            {
                if (token != null)
                {
                    bool shouldAdd = true;
                    for (var index = 0; index < mergedTokens.Count && shouldAdd; index++)
                    {
                        // It is included in one of the current tokens
                        if (token.Start >= mergedTokens[index].Start && token.End <= mergedTokens[index].End)
                        {
                            shouldAdd = false;
                        }

                        // If it contains overlaps
                        if (token.Start > mergedTokens[index].Start && token.Start < mergedTokens[index].End)
                        {
                            shouldAdd = false;
                        }

                        // It includes one of the tokens and should replace the included one
                        if (token.Start <= mergedTokens[index].Start && token.End >= mergedTokens[index].End)
                        {
                            shouldAdd = false;
                            mergedTokens[index] = token;
                        }
                    }

                    if (shouldAdd)
                    {
                        mergedTokens.Add(token);
                    }
                }
            }

            foreach (var token in mergedTokens)
            {
                var start = token.Start;
                var length = token.Length;
                var substr = text.Substring(start, length);

                var er = new ExtractResult
                {
                    Start = start,
                    Length = length,
                    Text = substr,
                    Type = extractorName,
                    Data = null,
                    Metadata = token.Metadata,
                };

                ret.Add(er);
            }

            return ret;
        }

        public Token Clone()
        {
            return (Token)MemberwiseClone();
        }
    }
}

File: Java/libraries/recognizers-text-date-time/src/main/java/com/microsoft/recognizers/text/datetime/utilities/Token.java
// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.

package com.microsoft.recognizers.text.datetime.utilities;

import com.microsoft.recognizers.text.ExtractResult;
import com.microsoft.recognizers.text.Metadata;
import com.microsoft.recognizers.text.utilities.Match;
import com.microsoft.recognizers.text.utilities.RegExpUtility;

import java.util.ArrayList;
import java.util.Collection;
import java.util.List;
import java.util.regex.Pattern;

public class Token {
    private final int start;
    private final int end;
    private final Metadata metadata;

    public Token(int start, int end, Metadata metadata) {
        this.start = start;
        this.end = end;
        this.metadata = metadata;
    }

    public Token(int start, int end) {
        this.start = start;
        this.end = end;
        this.metadata = null;
    }

    public int getStart() {
        return start;
    }

    public int getEnd() {
        return end;
    }

    public int getLength() {
        return end < start ? 0 : end - start;
    }

    public static List<ExtractResult> mergeAllTokens(List<Token> tokens, String text, String extractorName) {
        List<ExtractResult> result = new ArrayList<>();
        List<Token> mergedTokens = new ArrayList<>();

        tokens.sort((o1, o2) -> {
            if (o1.start != o2.start) {
                return o1.start - o2.start;
            }

            return o2.getLength() - o1.getLength();
        });

        for (Token token : tokens) {
            if (token != null) {
                boolean bAdd = true;
                for (int i = 0; i < mergedTokens.size() && bAdd; i++) {
                    // It is included in one of the current tokens
                    if (token.start >= mergedTokens.get(i).start && token.end <= mergedTokens.get(i).end) {
                        bAdd = false;
                    }

                    // If it contains overlaps
                    if (token.start > mergedTokens.get(i).start && token.start < mergedTokens.get(i).end) {
                        bAdd = false;
                    }

                    // It includes one of the tokens and should replace the included one
                    if (token.start <= mergedTokens.get(i).start && token.end >= mergedTokens.get(i).end) {
                        bAdd = false;
                        mergedTokens.set(i, token);
                    }
                }

                if (bAdd) {
                    mergedTokens.add(token);
                }
            }
        }

        for (Token token : mergedTokens) {
            String substring = text.substring(token.start, token.end);

            ExtractResult er = new ExtractResult(token.start, token.getLength(), substring, extractorName, null, token.metadata);

            result.add(er);
        }

        return result;
    }

    public static Collection<Token> getTokenFromRegex(Pattern pattern, String text) {
        Collection<Token> result = new ArrayList<>();

        for (Match match : RegExpUtility.getMatches(pattern, text)) {
            result.add(new Token(match.index, match.index + match.length));
        }

        return result;
    }
}
